### PR TITLE
Changed banned TSV thread behavior, refactored modmail archiving

### PR DESCRIPTION
**Pull request by not-an-aardvark**
Now we have a Reddit.js helper function to get entire listings of things, so we don't need to deal with pagination for search, modmails, etc. I also refactored the reddit search function to allow for different types of searches in case we need them in the future.

Also changed the banned TSV thread policy. Currently the banned linkflair has text '[Banned User] Trainer Shiny Value' and CSS class 'banned'.

### CODE REVIEW

File: api/controllers/UserController.js [hunk 1]
```diff
@@ -283,7 +283,7 @@ module.exports = {
         user: req.user.name,
         type: "banUser",
         content: "Banned /u/" + req.params.username
-      });
+      }).exec(function () {});
```
Comment (YaManicKill) [inline api/controllers/UserController.js]: This, I guess, is a bug fix for 2.2, so add this as a separate PR to 2.2, and we'll get that in.

File: api/controllers/UserController.js [hunk 2]
```diff
@@ -268,12 +268,12 @@ module.exports = {
...
         promises.push(Ban.addUsernote(req.user.redToken, req.user.name, 'pokemontrades', req.params.username, req.params.banNote));
         promises.push(Ban.addUsernote(req.user.redToken, req.user.name, 'SVExchange', req.params.username, req.params.banNote));
-        promises.push(Ban.removeTSVThreads(req.user.redToken, req.params.username));
+        promises.push(Ban.markTSVThreads(req.user.redToken, req.params.username));
         promises.push(Ban.updateBanlist(req.user.redToken, req.params.username, req.params.banlistEntry, unique_fcs, igns, req.params.knownAlt));
         promises.push(Ban.localBanUser(req.params.username));
       }
       Promise.all(promises).then(function () {
-        console.log('Process to ban /u/' + req.params.username + 'was completed successfully.');
+        console.log('Process to ban /u/' + req.params.username + ' was completed successfully.');
```
Comment (YaManicKill) [inline api/controllers/UserController.js]: This too.

File: api/services/Reddit.js [hunk 3]
```diff
@@ -34,7 +35,7 @@ exports.refreshToken = async function(refreshToken) {
   }
 };
 
-var makeRequest = async function (refreshToken, requestType, url, data, rateLimitRemainingThreshold) {
+var makeRequest = async function (refreshToken, requestType, url, data, rateLimitRemainingThreshold, silentErrors) {
```
Comment (YaManicKill) [inline api/services/Reddit.js]: Going to have to explain this a bit.

File: api/services/Reddit.js [hunk 4]
```diff
@@ -79,13 +94,19 @@ exports.getBothFlairs = async function (refreshToken, user) {
...
 
-exports.setFlair = function (refreshToken, name, cssClass, text, subreddit) {
+exports.setUserFlair = function (refreshToken, name, cssClass, text, subreddit) {
   var actual_sub = sails.config.debug.reddit ? sails.config.debug.subreddit : subreddit;
   var url = 'https://oauth.reddit.com/r/' + actual_sub + '/api/flair';
   var data = {api_type: 'json', css_class: cssClass, name: name, text: text};
   return makeRequest(refreshToken, 'POST', url, data, 5);
 };
 
+exports.setLinkFlair = function (refreshToken, subreddit, link_id, cssClass, text) {
```
Comment (YaManicKill) [inline api/services/Reddit.js]: (I'm not suggesting changing the name of this, because it is how reddit calls it but....)

I really hate how it's called link flair, as we don't have link posts, only self posts. I wish they called it post flair.

File: api/services/Reddit.js (hunk 3 above)
Comment (not-an-aardvark) [inline api/services/Reddit.js]: When we attempt to lock an archived post, Reddit returns a 400 error. We can test for that case somewhere, but we don't need it to be logged as an error message, since in this case the 400 response is expected. So we turn on the silentErrors flag to squelch the error message.

File: api/controllers/UserController.js (hunk 1 above)
Comment (not-an-aardvark) [inline api/controllers/UserController.js]: [Actually it works in 2.2](https://github.com/YaManicKill/flairhq/blob/version-2.2/api/controllers/UserController.js#L298). I think I got rid of it while refactoring stuff for 2.3.

File: api/controllers/UserController.js (hunk 2 above)
Comment (not-an-aardvark) [inline api/controllers/UserController.js]: The completion message was added in a [2.3 refactor](https://github.com/YaManicKill/flairhq/blame/version-2.3/api/controllers/UserController.js#L276) -- it doesn't exist in 2.2 at all.

File: api/controllers/UserController.js (hunk 1 above)
Comment (YaManicKill) [inline api/controllers/UserController.js]: Ah, righty. Makes sense.

File: api/controllers/UserController.js (hunk 2 above)
Comment (YaManicKill) [inline api/controllers/UserController.js]: I wouldn't have said to put it in a 2.2 PR if it was the only 2.2 fix anyway, so yeah :-P

File: api/services/Reddit.js (hunk 3 above)
Comment (YaManicKill) [inline api/services/Reddit.js]: Ah right. Can we call it "silenceErrors" rather than "silentErrors" as the latter implies that we are passing an error through itself.

File: api/services/Reddit.js [hunk 5]
```diff
@@ -50,9 +51,11 @@ var makeRequest = async function (refreshToken, requestType, url, data, rateLimi
...
   };
   let response = await request(options).catch(function (error) {
-    console.log('Reddit error: ' + requestType + ' request sent to ' + url + ' returned ' + error.statusCode +
-      ' - ' + error.statusMessage + '\nForm data sent: ' + JSON.stringify(data));
-    throw {statusCode: 502, error: 'Reddit responded with status code ' + error.statusCode};
+    if (!silentErrors) {
+      console.log('Reddit error: ' + requestType + ' request sent to ' + url + ' returned ' + error.statusCode);
+      console.log('Form data sent: ' + JSON.stringify(data));
+    }
+    throw {statusCode: error.statusCode, error: '(Reddit response)'};
```
Comment (YaManicKill) [inline api/services/Reddit.js]: Doesn't this want to be inside the if statement?

File: api/services/Reddit.js [hunk 6]
```diff
@@ -65,6 +68,18 @@ var makeRequest = async function (refreshToken, requestType, url, data, rateLimi
   return bodyJson;
 };
 
+var getEntireListing = async function (refreshToken, endpoint, query, rateThreshold, after, before) {
+  var url = endpoint + query + (query ? '&' : '?') + 'count=102&limit=100' + (after ? '&after=' + after : '') + (before ? '&before=' + before : '');
+  var batch = await makeRequest(refreshToken, 'GET', url, undefined, rateThreshold, after, before);
+  var results = batch.data.children;
+  if ((after || after === undefined) && batch.data.after === null || before && batch.data.before === null) {
```
Comment (YaManicKill) [inline api/services/Reddit.js]: I think there might be something wrong with this if statement?

If after is truthy, or it's undefined? 

Are the brackets wrong? Run throguh the logic with me.

File: api/services/Reddit.js (hunk 5 above)
Comment (not-an-aardvark) [inline api/services/Reddit.js]: I think we should still throw the error when `silenceErrors` is on, so that it's clear that an error occured. We can handle the error further down.

File: api/services/Reddit.js (hunk 6 above)
Comment (not-an-aardvark) [inline api/services/Reddit.js]: You're right, I think there might have been a logic error, so I simplified the logic a bit. It's still a bit complicated, so I'll explain the process.

`getEntireListing()` can be called in one of the following three states:
Case 1. `after` is defined and `before` is not (i.e. we're searching forward, getting all listings after a certain point)
Case 2. `before` is defined and `after` is not (i.e. we're searching backward, getting all listings before a certain point)
Case 3. `after` and `before` are both `undefined` (i.e. we're searching forward, getting the entire listing starting at the top)

Whenever we make a call, reddit returns `batch.data.after` and `batch.data.before`, which are intended to be starting points for the next search. However, one of these will be `null` if there are no more results to get. We want to recursively call `getEntireListing()` again with these new parameters as long as they're not `null`.

Example: Suppose there are 250 modmails in total, and we want to get all of them. 

1. We call `getEntireListing()` with `after` and `before` both undefined. Reddit returns a `batch` containing modmails 1-100. It also sends `batch.data.after=<id of 100th modmail>` and `batch.data.before=null`.
2. We recursively call `getEntireListing()` again, with `after=<id of 100th modmail>` and `before=undefined`. Reddit returns the batch containing modmails 101-200. It also sends `batch.data.after=<id of 200th modmail>` and `batch.data.before=<id of 101st modmail>`.
3. We recursively call `getEntireListing()` again, with `after=<id of 200th modmail>` and `before=undefined`. Reddit returns the batch containing modmails 201-250. It also sends `batch.data.after=null` (since there are no more results to get) and `batch.data.before=<id of 201st modmail>`.
4. We notice that `batch.data.after` is `null`, so we conclude that the search has finished and we return the final list.

After we get the batch, we do the following:

    after = before ? undefined : batch.data.after;
    before = before ? batch.data.before : undefined;

In other words:

* If `before` is undefined (i.e. we're searching forward), set `after` to `batch.data.after` and keep `before` as `undefined` for the next search.
* If `before` is defined (i.e. we're searching backward), set `after` to `undefined` and set `before` to `batch.data.before` for the next search.

Afterwards, test `if (!before && !after)`. One of these will always be undefined, and the other will be the relevant `batch.data` response from reddit. If that `batch.data` response is `null`, then the search is over; the `if` statement will check out and the results will be returned. If the `batch.data` response is something else, the if statement will not check out, and a query will be made with the new parameters.

File: api/services/Reddit.js (hunk 3 above)
Comment (not-an-aardvark) [inline api/services/Reddit.js]: Fixed.

**Comment by YaManicKill** [comment]
This needs rebased again. sorry.
